### PR TITLE
zero the gradients after updating weights

### PR DESCRIPTION
**Pull request by bintang-aswam**
Manually zero the gradients after updating weights by using machine epsilon for standard float (64-bit).

Fixes #ISSUE_NUMBER

## Checklist

- [ ] The issue that is being fixed is referred in the description (see above "Fixes #ISSUE_NUMBER")
- [ ] Only one issue is addressed in this pull request
- [ ] Labels from the issue that this PR is fixing are added to this pull request
- [ ] No unnecessary issues are included into this pull request.

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
Would you be able to share details on why do you want to do this change.
Setting them to None is equivalent to setting them to exactly 0. which is best no?

**Comment by bintang-aswam** [comment]
> Would you be able to share details on why do you want to do this change. Setting them to None is equivalent to setting them to exactly 0. which is best no?
answer:
@albanD, In other words, setting them to `weight.grad = None` is equivalent to setting them to `weight.grad.zero_()`. However, there are few conceptual difference in the context of numerically zeroing between **zeroing the gradients** vs **tensor-multiplication of loss by machine-epsilon** as follows:
| Indicator                 | Tensor-Multiplication of Loss by Machine Epsilon                                          | Zeroing the Gradients                                                              |
|---------------------------|-------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------|
| Goal                      | to scale the loss to near-zero values                                                     | to clear up the accumulated gradients before the next backward pass                |
| Effect on Backpropagation | to generate gradients extremely small                                                     | to ensure gradients from previous steps don't interfere                            |
| Typical Use Case          | used to simulate vanishing gradients                                                      | mostly used for standard practice within training loops                            |
| Numerical Impact          | to test sensitivity of model to small perturbations including to test numerical stability | just reset gradients to None or literally zero (which is numerically meaningless). |

**Comment by bintang-aswam** [comment]
> Would you be able to share details on why do you want to do this change. Setting them to None is equivalent to setting them to exactly 0. which is best no?

answer:
@albanD, In other words, setting them to `weight.grad = None` is equivalent to setting them to `weight.grad.zero_()`. However, there are few conceptual difference in the context of numerically zeroing between **zeroing the gradients** vs **tensor-multiplication of loss by machine-epsilon** as follows:
image

**Comment by albanD** [comment]
Thanks for sharing these details.
I'm still not sure why this particular tutorial should be updated?
Maybe you want to do momentum-like updates here? If so, we can add a momentum term there?
